File: homeworks/homework5.1/src/main/java/org/example/MegaDataPool.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class MegaDataPool {
    private final List<MegaData> dataPool = new ArrayList<>();
    private final List<Boolean> isDataUsed = new ArrayList<>();

    public MegaDataPool(int size) {
        for (int i = 0; i < size; i++) {
            dataPool.add(new MegaData());
            isDataUsed.add(false);
        }
    }

    public MegaData acquire() {
        for (int i = 0; i < isDataUsed.size(); i++) {
            if (!isDataUsed.get(i)) {
                isDataUsed.set(i, true);
                return dataPool.get(i);
            }
        }
        return null;
    }

    public void release(MegaData obj) {
        int index = dataPool.indexOf(obj);
        if (index != -1 && isDataUsed.get(index)) {
            isDataUsed.set(index, false);
            obj.reset();
        }
    }

    public int getSize() {
        return dataPool.size();
    }

    public int getUsedCount() {
        return (int) isDataUsed.stream().filter(b -> b).count();
    }
}

File: practices/practice5.1/src/MegaDataPool.cpp
#include "MegaDataPool.hpp"
#include <algorithm>
#include <sstream>

MegaDataPool::MegaDataPool(size_t size) : dataPool(size), isDataUsed(size, false) {}

MegaData& MegaDataPool::acquire() {
    for (size_t i = 0; i < dataPool.size(); ++i) {
        if (!isDataUsed[i]) {
            isDataUsed[i] = true;
            return dataPool[i];
        }
    }
    throw std::runtime_error("No available data objects in the pool");
}

void MegaDataPool::release(MegaData& obj) {
    for (size_t i = 0; i < dataPool.size(); ++i) {
        if (&dataPool[i] == &obj) {
            isDataUsed[i] = false;
            return;
        }
    }
}

size_t MegaDataPool::getSize() const {
    return dataPool.size();
}

size_t MegaDataPool::getUsedCount() const {
    return std::ranges::count(isDataUsed, true);
}
